### PR TITLE
IfcfgUtil: Remediate `connection_seems_active()` for controller

### PR DESCRIPTION
**Pull request by liangwen12year**
For the active bonding/bridge connection, when all the port connection
profiles are brought down firstly, then the operstate of the controller
device will change into "down" automatically, which denotes missing the
carrier for the controller device. However, the connection for the
controller device should still be considered as active as long as
there is any valid IP address configured.

Signed-off-by: Wen Liang <liangwen12year@gmail.com>

### CODE REVIEW

File: library/network_connections.py [hunk 1]
```diff
@@ -683,7 +683,21 @@ def connection_seems_active(cls, name):
             return None
 
         if i_content.strip() != "up":
-            return False
+            # For the active bonding/bridge connection, when all the port connection
+            # profiles are brought down firstly, then the operstate of the controller
+            # device will change into "down" automatically, which denotes missing the
+            # carrier for the controller device. However, the connection for the
+            # controller device stays active in such a situation.
```
Comment (nhosoi) [inline library/network_connections.py:690]: Thank you, @liangwen12year. The serialized test passed with your patch.
Let me just confirm this is the final solution. I somehow thought you were going to clean up the device (and connection?) completely in the bond tests. Your comment looks we allow the odd state instead... Is it ok?

**Review by nhosoi — DISMISSED** [review]
Thank you for the detailed explanation, @liangwen12year.
lgtm

**Comment by richm** [comment]
The rhel-x/ansible-2.12 test is hung in CI at tests_bond_removal_nm.yml
```
TASK [linux-system-roles.network : Configure networking connection profiles] ***
task path: /tmp/tmp3bnjgjfp/tests/roles/linux-system-roles.network/tasks/main.yml:75
Wednesday 25 May 2022  00:05:52 +0000 (0:00:00.032)       0:00:28.863 ********* 
```
here are the last few lines from the guest journal
```
[   40.664931] nm-bond: (slave test1): Enslaving as an active interface with an up link
[   40.667412] testbr: port 1(test1p) entered blocking state
[   40.669039] testbr: port 1(test1p) entered forwarding state
[   40.670680] IPv6: ADDRCONF(NETDEV_CHANGE): nm-bond: link becomes ready
[   40.699717] testbr: port 2(test2p) entered disabled state
[   40.701756] nm-bond: (slave test2): Enslaving as a backup interface with an up link
[   40.704283] testbr: port 2(test2p) entered blocking state
[   40.705839] testbr: port 2(test2p) entered forwarding state
[31283.731087] systemd-journald[594]: Data hash table of /run/log/journal/0ce9f85c1a1a4c2f884f096977a4602e/system.journal has a fill level at 75.0 (6051 of 8064 items, 4644864 file size, 767 bytes per hash table item), suggesting rotation.
[31283.739327] systemd-journald[594]: /run/log/journal/0ce9f85c1a1a4c2f884f096977a4602e/system.journal: Journal header limits reached or header out-of-date, rotating.
```
Are you able to run this test locally using `tox -e qemu-ansible-core-2.12 -- image-name rhel-x tests/tests_bond_removal_nm.yml`?

**Comment by richm** [comment]
here is the output of `ip a` from the rhel9 vm:
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:54:00:12:34:56 brd ff:ff:ff:ff:ff:ff
    altname enp0s3
    altname ens3
    inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic noprefixroute eth0
       valid_lft 28247sec preferred_lft 28247sec
    inet6 fec0::5054:ff:fe12:3456/64 scope site dynamic mngtmpaddr 
       valid_lft 86287sec preferred_lft 14287sec
    inet6 fe80::5054:ff:fe12:3456/64 scope link 
       valid_lft forever preferred_lft forever
3: test1p@test1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master testbr state UP group default qlen 1000
    link/ether 96:8a:0a:b4:c0:8b brd ff:ff:ff:ff:ff:ff
    inet6 fe80::948a:aff:feb4:c08b/64 scope link 
       valid_lft forever preferred_lft forever
4: test1@test1p: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc noqueue master nm-bond state UP group default qlen 1000
    link/ether da:e4:d4:a4:b1:bf brd ff:ff:ff:ff:ff:ff
5: test2p@test2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master testbr state UP group default qlen 1000
    link/ether 2e:a9:99:5b:7f:83 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::2ca9:99ff:fe5b:7f83/64 scope link 
       valid_lft forever preferred_lft forever
6: test2@test2p: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc noqueue master nm-bond state UP group default qlen 1000
    link/ether da:e4:d4:a4:b1:bf brd ff:ff:ff:ff:ff:ff
7: testbr: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 2e:a9:99:5b:7f:83 brd ff:ff:ff:ff:ff:ff
    inet 192.0.2.1/24 scope global testbr
       valid_lft forever preferred_lft forever
    inet6 2001:db8::2ca9:99ff:fe5b:7f83/64 scope global dynamic mngtmpaddr 
       valid_lft 1616sec preferred_lft 1616sec
    inet6 2001:db8::1/32 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::20f7:37ff:feb9:5bf9/64 scope link 
       valid_lft forever preferred_lft forever
9: nm-bond: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether da:e4:d4:a4:b1:bf brd ff:ff:ff:ff:ff:ff
    inet 192.0.2.223/24 brd 192.0.2.255 scope global dynamic noprefixroute nm-bond
       valid_lft 72sec preferred_lft 72sec
    inet6 2001:db8::789a:15fa:6101:d6c1/64 scope global dynamic noprefixroute 
       valid_lft 1617sec preferred_lft 1617sec
    inet6 fe80::7ab8:d8e1:ef99:203/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
```

**Comment by nhosoi** [comment]
> The rhel-x/ansible-2.12 test is hung in CI at tests_bond_removal_nm.yml

I have no problem to run this command line on my fedora 35... It's successfully finished.
```
tox -e qemu-ansible-core-2.12 -- --image-name rhel-x tests/tests_bond_removal_nm.yml
```
Here's the nm-bond info on the managed host.
```
# cat /etc/redhat-release 
Red Hat Enterprise Linux release 9.0 Beta (Plow)

# ip address show nm-bond
9: nm-bond: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 52:fc:ea:75:05:fd brd ff:ff:ff:ff:ff:ff
    inet 192.0.2.122/24 brd 192.0.2.255 scope global dynamic noprefixroute nm-bond
       valid_lft 56sec preferred_lft 56sec
    inet6 2001:db8::1f2/128 scope global dynamic noprefixroute 
       valid_lft 57sec preferred_lft 57sec
    inet6 2001:db8::c42f:cf37:aeb8:640e/64 scope global dynamic noprefixroute 
       valid_lft 1621sec preferred_lft 1621sec
    inet6 fe80::12ad:fa8d:c440:c422/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
```

**Comment by richm** [comment]
Here are the processes running in the rhel9 vm:
```
root         733       1  0 May24 ?        00:00:02 /usr/sbin/NetworkManager --no-daemon
dnsmasq     8574       1  0 May24 ?        00:00:00 dnsmasq --pid-file=/run/dhcp_testbr.pid --dhcp-leasefile=/run/dhcp_testbr.lease --dhcp-range=192.0.2.1,192.0.2.254,240 --dhcp-range=2001:DB8::10,2001:DB8::1FF,slaac,64,240 --enable-ra --interface=testbr --bind-interfaces
root       14857    1389  0 May24 pts/0    00:00:00 /bin/sh -c /usr/bin/python3 /root/.ansible/tmp/ansible-tmp-1653437152.314268-236772-257593223212697/AnsiballZ_network_connections.py && sleep 0
root       14871   14857  0 May24 pts/0    00:00:00 /usr/bin/python3 /root/.ansible/tmp/ansible-tmp-1653437152.314268-236772-257593223212697/AnsiballZ_network_connections.py
```
Here are the arguments passed to the network_connections module:
```
May 24 20:05:52 ibm-p8-kvm-03-guest-02.virt.pnr.lab.eng.rdu2.redhat.com python3[14871]: ansible-network_connections Invoked with provider=nm connections=[{'name': 'bond0', 'state': 'up', 'type': 'bond', 'interface_name': 'nm-bond', 'bond': {'mode': 'active-backup', 'miimon': 110}}, {'name': 'bond0.0', 'state': 'up', 'type': 'ethernet', 'interface_name': 'test1', 'controller': 'bond0'}, {'name': 'bond0.1', 'state': 'up', 'type': 'ethernet', 'interface_name': 'test2', 'controller': 'bond0'}] __header=#
                                                                                        # Ansible managed
                                                                                        #
                                                                                         ignore_errors=False force_state_change=False __debug_flags=
```

**Comment by richm** [comment]
> > The rhel-x/ansible-2.12 test is hung in CI at tests_bond_removal_nm.yml
> 
> I have no problem to run this command line on my fedora 35... It's successfully finished.
> 
> ```
> tox -e qemu-ansible-core-2.12 -- --image-name rhel-x tests/tests_bond_removal_nm.yml
> ```
> 
> Here's the nm-bond info on the managed host.
> 
> ```
> # cat /etc/redhat-release 
> Red Hat Enterprise Linux release 9.0 Beta (Plow)
> 
> # ip address show nm-bond
> 9: nm-bond: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
>     link/ether 52:fc:ea:75:05:fd brd ff:ff:ff:ff:ff:ff
>     inet 192.0.2.122/24 brd 192.0.2.255 scope global dynamic noprefixroute nm-bond
>        valid_lft 56sec preferred_lft 56sec
>     inet6 2001:db8::1f2/128 scope global dynamic noprefixroute 
>        valid_lft 57sec preferred_lft 57sec
>     inet6 2001:db8::c42f:cf37:aeb8:640e/64 scope global dynamic noprefixroute 
>        valid_lft 1621sec preferred_lft 1621sec
>     inet6 fe80::12ad:fa8d:c440:c422/64 scope link noprefixroute 
>        valid_lft forever preferred_lft forever
> ```

Then probably some previous test didn't clean up properly, which made this test hang.

**Comment by richm** [comment]
> > > The rhel-x/ansible-2.12 test is hung in CI at tests_bond_removal_nm.yml
> > 
> > 
> > I have no problem to run this command line on my fedora 35... It's successfully finished.
> > ```
> > tox -e qemu-ansible-core-2.12 -- --image-name rhel-x tests/tests_bond_removal_nm.yml
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Here's the nm-bond info on the managed host.
> > ```
> > # cat /etc/redhat-release 
> > Red Hat Enterprise Linux release 9.0 Beta (Plow)
> > 
> > # ip address show nm-bond
> > 9: nm-bond: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
> >     link/ether 52:fc:ea:75:05:fd brd ff:ff:ff:ff:ff:ff
> >     inet 192.0.2.122/24 brd 192.0.2.255 scope global dynamic noprefixroute nm-bond
> >        valid_lft 56sec preferred_lft 56sec
> >     inet6 2001:db8::1f2/128 scope global dynamic noprefixroute 
> >        valid_lft 57sec preferred_lft 57sec
> >     inet6 2001:db8::c42f:cf37:aeb8:640e/64 scope global dynamic noprefixroute 
> >        valid_lft 1621sec preferred_lft 1621sec
> >     inet6 fe80::12ad:fa8d:c440:c422/64 scope link noprefixroute 
> >        valid_lft forever preferred_lft forever
> > ```
> 
> Then probably some previous test didn't clean up properly, which made this test hang.

. . . but I'm not sure what, since this isn't using batch mode . . .

**Comment by nhosoi** [comment]
> . . . but I'm not sure what, since this isn't using batch mode . . .

Right. That's what I was going to ask... :) Need to find out how to reproduce it...

**Comment by nhosoi** [comment]
It looks the hang in rhel-x was a fluke???

Thanks for waiting, @liangwen12year. Please go ahead and push the tox fix...

**Comment by richm** [comment]
There are two python issues which need to be addressed
```
ERROR: Found 2 validate-modules issue(s) which need to be resolved:
ERROR: plugins/modules/network_connections.py:697:27: use-run-command-not-popen: subprocess.Popen call found. Should be module.run_command
ERROR: plugins/modules/network_connections.py:698:30: use-run-command-not-popen: subprocess.Popen call found. Should be module.run_command
See documentation for help: https://docs.ansible.com/ansible-core/2.12/dev_guide/testing/sanity/validate-modules.html
```

**Comment by nhosoi** [comment]
@liangwen12year, unforunately, your new patch makes the bond removal tests fail again. I double-checked the previous patch was ok.

Here's the difference:
```
--- /usr/share/ansible/roles/rhel-system-roles.network/library/network_connections.py	2022-05-24 20:08:55.000000000 -0400
+++ network_connections.py	2022-05-26 12:02:32.725220317 -0400
@@ -658,9 +658,9 @@
                 with open(path, "w") as text_file:
                     text_file.write(h)
 
     @classmethod
-    def connection_seems_active(cls, name):
+    def connection_seems_active(cls, name, run_env):
         # we don't know whether a ifcfg file is currently active,
         # and we also don't know which.
         #
         # Do a very basic guess based on whether the interface
@@ -687,11 +687,11 @@
             # profiles are brought down firstly, then the operstate of the controller
             # device will change into "down" automatically, which denotes missing the
             # carrier for the controller device. However, the connection for the
             # controller device stays active in such a situation.
-            ip_address = subprocess.check_output(
+            ip_address = run_env.run_command(
                 ["ip", "address", "show", content["DEVICE"]]
-            ).decode("UTF-8")
+            )
             if "inet" in ip_address:
                 if ip_address.count("inet") == 1 and "inet6 fe80" in ip_address:
                     return False
                 else:
@@ -2627,9 +2627,9 @@
                     idx, "ifcfg file '%s' does not exist%s" % (path, in_checkmode)
                 )
             return
 
-        is_active = IfcfgUtil.connection_seems_active(name)
+        is_active = IfcfgUtil.connection_seems_active(name, self.run_env)
         is_modified = self.connection_modified_earlier(idx)
         force_state_change = self.connection_force_state_change(connection)
```

**Comment by nhosoi** [comment]
Snippet of the error output from `network/tests_bond_removal_initscripts.yml`:
```
TASK [rhel-system-roles.network : Configure networking connection profiles] ****
task path: /usr/share/ansible/roles/rhel-system-roles.network/tasks/main.yml:75
Thursday 26 May 2022  11:47:33 -0400 (0:00:00.484)       0:00:16.620 ********** 
[007] <info>  #0, state:up persistent_state:present, 'bond0': add ifcfg-rh profile 'bond0'
[008] <info>  #1, state:up persistent_state:present, 'bond0.0': add ifcfg-rh profile 'bond0.0'
[009] <info>  #2, state:up persistent_state:present, 'bond0.1': add ifcfg-rh profile 'bond0.1'
[010] <info>  #0, state:up persistent_state:present, 'bond0': up connection bond0 (not-active)
[011] <info>  #0, state:up persistent_state:present, 'bond0': call 'ifup bond0': rc=1, out='b'^LDetermining IP information for nm-bond... failed.
'', err='b"WARN      : [ifup] You are using 'ifup' script provided by 'network-scripts', which are now deprecated.
WARN      : [ifup] 'network-scripts' will be removed in one of the next major releases of RHEL.
WARN      : [ifup] It is advised to switch to 'NetworkManager' instead - it provides 'ifup/ifdown' scripts as well.
WARN      : [ifup] You are using 'ifup' script provided by 'network-scripts', which are now deprecated.
WARN      : [ifup] 'network-scripts' will be removed in one of the next major releases of RHEL.
WARN      : [ifup] It is advised to switch to 'NetworkManager' instead - it provides 'ifup/ifdown' scripts as well.
WARN      : [ifup] You are using 'ifup' script provided by 'network-scripts', which are now deprecated.
WARN      : [ifup] 'network-scripts' will be removed in one of the next major releases of RHEL.
WARN      : [ifup] It is advised to switch to 'NetworkManager' instead - it provides 'ifup/ifdown' scripts as well.
dhclient(22477) is already running - exiting.

This version of ISC DHCP is based on the release available
on ftp.isc.org. Features have been added and other changes
have been made to the base software release in order to make
it work better with this distribution.

Please report issues with this software via:
https://bugzilla.redhat.com/

exiting.
"'
[012] <error> #0, state:up persistent_state:present, 'bond0': call 'ifup bond0' failed with exit status 1

MSG:

error: call 'ifup bond0' failed with exit status 1
```

**Comment by nhosoi** [comment]
Great! Thank you, @richm . The failure was fixed by your patch.

**Comment by richm** [comment]
I pushed another patch to https://github.com/liangwen12year/network/commit/91ed24b32c0cf09d8fd6d98e2c3c0f22d98084ab#diff-6b3f5f6149ac1330c5e2c832286b63b3597158a998f2244ddbec2045ace1f794R691 to fix the ansible-test issue with tox / python 3.8

**Comment by nhosoi** [comment]
> I pushed another patch to [liangwen12year@91ed24b#diff-6b3f5f6149ac1330c5e2c832286b63b3597158a998f2244ddbec2045ace1f794R691](https://github.com/liangwen12year/network/commit/91ed24b32c0cf09d8fd6d98e2c3c0f22d98084ab#diff-6b3f5f6149ac1330c5e2c832286b63b3597158a998f2244ddbec2045ace1f794R691) to fix the ansible-test issue with tox / python 3.8

Yes, my test used your latest patch, which passed the test.

@liangwen12year, please update this pr with the latest commit by @richm. Thanks!

File: library/network_connections.py [hunk 2]
```diff
@@ -683,7 +683,22 @@ def connection_seems_active(cls, name):
...
+            _unused_, out, _unused_ = run_env.run_command(
+                ["ip", "address", "show", content["DEVICE"]],
+                "utf-8",
+            )
```
Comment (thom311) [inline library/network_connections.py:694]: what exceptions can `run_command` raise? I think, any exception should be silently swallowed.

**Comment by nhosoi** [comment]
> [liangwen12year](https://github.com/liangwen12year) [force-pushed](https://github.com/linux-system-roles/network/compare/bf3b5d583ef8d7b56ff25ecbe269f278a274b983..c872dae5c4b6f26b392cc9548e7a35ae569e42fd) the bond_fix branch from[ bf3b5d5 ](https://github.com/linux-system-roles/network/commit/bf3b5d583ef8d7b56ff25ecbe269f278a274b983)to c872dae

Bond tests were successful.

File: library/network_connections.py (hunk 2 above)
Comment (liangwen12year) [reply library/network_connections.py]: Updated.

**Comment by nhosoi** [comment]
Thank you, @liangwen12year, @richm. This patch fixes the serialized bond test issue.

I'd like to have a confirmation on one more test failure in tests_infiniband_nm.yml.
```
TASK [rhel-system-roles.network : Configure networking connection profiles] ****
task path: /usr/share/ansible/roles/rhel-system-roles.network/tasks/main.yml:75
<<snip>>
[001] <error> #0, state:None persistent_state:present, 'ib0': profile specifies interface_name 'ib0' but no such infiniband interface exists
MSG:
error: profile specifies interface_name 'ib0' but no such infiniband interface exists
```
Have you seen this failure? It looks it's another error caused by the serialization, doesn't it?

**Review by nhosoi — APPROVED** [review]
lgtm

**Comment by thom311** [comment]
I don't have a strong opinion about this. After all, we need some sort of heuristic to determine whether the device is up, so we know when to call ifdown. That heuristic is necessarily not perfect, but this seems a sensible improvement. +1.

**Pull request merged**